streams/src/test/java/org/apache/kafka/streams/integration/AbstractJoinIntegrationTest.java

```diff
@@ -0,0 +1,290 @@
+
+        try {
+            streams.start();
+
+            long ts = System.currentTimeMillis();
+
+            final Iterator<List<String>> resultIterator = expectedResult.iterator();
+            for (final Input<String> singleInput : input) {
+                producer.send(new ProducerRecord<>(singleInput.topic, null, ++ts, singleInput.record.key, singleInput.record.value)).get();
+
+                List<String> expected = resultIterator.next();
+
+                if (expected != null) {
+                    checkResult(OUTPUT_TOPIC, expected);
+                    expectedFinalResult = expected.get(expected.size() - 1);
+                }
+            }
+
+            if (storeName != null)
+                checkQueryableStore(storeName, expectedFinalResult);
+
+        } finally {
+            streams.close();
+        }
+    }
+
+    /*
+     * Runs the actual test. Checks the final result only after expected number of records have been consumed.
+     */
+    void runTest(final String expectedFinalResult) throws Exception {
+        runTest(expectedFinalResult, null);
+    }
+
+    /*
+     * Runs the actual test. Checks the final result only after expected number of records have been consumed.
+     */
+    void runTest(final String expectedFinalResult, final String storeName) throws Exception {
+        IntegrationTestUtils.purgeLocalStreamsState(STREAMS_CONFIG);
+        streams = new KafkaStreams(builder.build(), new StreamsConfig(STREAMS_CONFIG));
+
+        try {
+            streams.start();
+
+            long ts = System.currentTimeMillis();
+
+            for (final Input<String> singleInput : input) {
+                producer.send(new ProducerRecord<>(singleInput.topic, null, ++ts, singleInput.record.key, singleInput.record.value)).get();
+            }
+
+            TestUtils.waitForCondition(new TestCondition() {
+                @Override
+                public boolean conditionMet() {
+                    return finalResultReached.get();
+                }
+            }, "Never received expected final result.");
+
+            checkResult(OUTPUT_TOPIC, expectedFinalResult, numRecordsExpected);
+
+            if (storeName != null)
+                checkQueryableStore(storeName, expectedFinalResult);
+
+        } finally {
+            streams.close();
+        }
+    }
+
+    /*
+     * Checks the embedded queryable state store snapshot
+     */
+    private void checkQueryableStore(final String queryableName, final String expectedFinalResult) {
+        final ReadOnlyKeyValueStore<Long, String> store = streams.store(queryableName, QueryableStoreTypes.<Long, String>keyValueStore());
+
+        final KeyValueIterator<Long, String> all = store.all();
+        final KeyValue<Long, String> onlyEntry = all.next();
+
+        assertThat(onlyEntry.key, is(anyUniqueKey));
+        assertThat(onlyEntry.value, is(expectedFinalResult));
+        assertThat(all.hasNext(), is(false));
+
+        all.close();
```

Comment:
this will never be called if one of the assertions fails

Reply:
Good point!